pkg/util/ipset/ipset.go

```diff
@@ -85,6 +87,39 @@ type IPSet struct {
 	MaxElem int
 	// PortRange specifies the port range of bitmap:port type ipset.
 	PortRange string
+	// TODO: add comment message for ipset
+}
+
+// Validate checks if a given ipset is valid or not.
+func (set *IPSet) Validate() bool {
```

Comment:
I'm a little confused on why you need this. It seems like Validate for an Entry does everything you need. Can you describe a chain of events where this function would return false?

Reply:
This function is called by 

```
func (runner *runner) CreateSet(set *IPSet, ignoreExistErr bool) error {
    ...
    // Validate ipset before creating
    valid := set.Validate()
    if !valid {
	return fmt.Errorf("Error creating ipset since it's invalid")
    }
    return runner.createSet(set, ignoreExistErr)
}
```

Possibly I need to change the PR title to `validate ipset set and entry - ipvs proxier validate entry before adding it and validate set before creating it`

Reply:
What I was asking was when would there be a situation where Validate() returns false? In other words, is there a set of consecutive events that would cause a set to be invalid? I can't think of any which is why I am confused as to why you even need it.

Reply:
Yes, three are...

```
// Check if protocol is valid for `HashIPPort`, `HashIPPortIP` and `HashIPPortNet` type set.
	if set.SetType == HashIPPort || set.SetType == HashIPPortIP || set.SetType == HashIPPortNet {
		if valid := validateHashFamily(set.HashFamily); !valid {
			return false
		}
	}
	// check set type
	if valid := validateIPSetType(set.SetType); !valid {
		return false
	}
	// check port range for bitmap type set
	if set.SetType == BitmapPort {
		if valid := validatePortRange(set.PortRange); !valid {
			return false
		}
	}
	// check hash size value of ipset
	if set.HashSize <= 0 {
		return false
	}
	// check max elem value of ipset
	if set.MaxElem <= 0 {
		glog.Errorf("Invalid maxelem value %d, should be >0", set.MaxElem)
		return false
	}
```

All situations above will return false by Validate().

I am not sure I FULLY understand...

Reply:
We should keep it in mind that this util pkg is public and can be used by other processes...

Reply:
I get that since it is a public pkg, you need it. However I am trying to understand at a deeper level why an IPSet would ever be invalid in the ipvs case

Reply:
>  However I am trying to understand at a deeper level why an IPSet would ever be invalid in the ipvs case

Ah, ipvs case will not cause error for now.

Reply:
Got it, thank you for clarifying :)